### PR TITLE
Fixed handling of pytrees in lbfgs

### PR DESCRIPTION
**Pull request by vroulet**
Fixed handling of pytrees by lbfgs, see https://github.com/google/jaxopt/pull/468#issuecomment-1632651231
Fixed previous test of inv_hessian_product and added test for two different representation of the inputs.

### CODE REVIEW

File: jaxopt/_src/lbfgs.py [hunk 1]
```diff
@@ -75,18 +47,18 @@ def inv_hessian_product(pytree: Any,
                         start: int = 0):
   """Product between an approximate Hessian inverse and a pytree.
 
-  Histories are pytrees of the same structure as `pytree`.
-  Leaves are arrays of shape `(history_size, ...)`, where
+  Histories are pytrees of the same structure as `pytree` execept
```
Comment (mblondel) [inline jaxopt/_src/lbfgs.py:50]:
```suggestion
  Histories are pytrees of the same structure as `pytree` except 
```

**Review by mblondel — APPROVED** [review]
LGTM apart from typo

File: tests/lbfgs_test.py [hunk 2]
```diff
@@ -62,14 +76,11 @@ def _inv_hessian_product(s, y, v, gamma=1.0):
   def body_fun(j, tup):
     rho, alpha, r = tup
     i = history_size - j - 1
-    # rho[i] = 1. / jnp.sum(s[i] * y[i])
```
Comment (mblondel) [inline tests/lbfgs_test.py:65]: Could you keep those comments please. I think they are useful.

**Comment by vroulet** [comment]
Rebased and squashed, ready to merge if it's ok with you. Thanks for the review!

**Review by mblondel — APPROVED** [review]

**Pull request merged**